Adding a tool HapCutToVcf.

nh13 opened the pull request:
This tool will convert the output of HapCut to a valid VCF, in a similar
fashion to how Picard's MergeBamAlignment post-processes the output of
an aligner.

tfenne commented on src/main/scala/com/fulcrumgenomics/cmdline/ClpGroups.scala:
```diff
@@ -58,11 +58,17 @@ object ClpGroups {
   final val Fastq     = classOf[_FastqGroup]
   final val Umi       = classOf[_Umi]
   final val Utilities = classOf[_UtilitiesGroup]
+  final val Phasing   = classOf[_PhasingGroup]
```
Are we likely to have > 1 tool for phasing over time?  If not, perhaps just a variants group?

nh13 replied:
Done

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+@clp(
+  description =
+    """
+      |Converts the output of HapCut to a VCF.
```
Maybe add a reference to HapCut here or at the end of the documentation?

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+      |
+      |The output of HAPCUT does not include all input variants, but simply those variants that are in phased blocks.
+      |This tool takes the original VCF and the output of HapCut, and produces two VCFs: (1) a VCF containing variants
+      |that were phased, and (2) a VCF containing variants that were not phased, such that all variants in the original
```
I'm guessing here, but is this because it's not possible to output unphased genotypes between phased genotypes in VCF?  I.e. if you have variants A-B-c-D with A, B and D phased, you can't stick `c` in the middle?  Otherwise it seems like it would be much nicer to have a single VCF with all variants, with phasing where possible.

nh13 replied:
I think the original intent was to have a VCF that was equivalent to what GATK's read-backed phasing produced, which did not produce A-B-c-D like you described, though it might well produce A-B-c-D where c was marked as not PASS, though I haven't checked.  Nonetheless, I think you are right to question the original intent and ask to just output a clean VCF with phasing information as the spec describes. Will do.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+      |that were phased, and (2) a VCF containing variants that were not phased, such that all variants in the original
+      |file are in one of two output VCFs.
+      |
+      |The VCF with phased variants will have the first variant of a haplotype block with a genotype "M\F", while the
```
This doesn't mean much to me - so perhaps a little more elaboration?  I'm guessing Mother and Father?  But we don't know parent, we just know what goes together.  Also is the `\` vs `|` VCF spec stuff?

tfenne replied:
Ok, having read a bunch more, I can't find anything in the spec about using `\` as a separator.  It says genotypes should always be specified with `|` or `/`.  What gives?

nh13 replied:
Good point, the VCF spec says that all phased variants should have `|`, and recommends to use the `PS` format field to with the coordinate first variant in the phasing set for all variants phased in one block.  

GATK, funnily enough, doesn't do this, instead it does what I describe above, which is the first (genomic coordinate) variant in a phase set is set to `\`, then the rest `|` (see "Note that the first heterozygous genotype ..." http://gatkforums.broadinstitute.org/gatk/discussion/45/purpose-and-operation-of-read-backed-phasing). 

What if I make it conform to the spec (all phased variants are `\` and `PS` is set correctly), but have an option to output "GATK-style" phased blocks?

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+  @arg(flag="i", doc="The output file from HAPCUT.") val input: FilePath,
+  @arg(flag="o", doc="The output VCF with variants that were phased by HAPCUT.") val output: PathToVcf,
+  @arg(flag="u", doc="The output VCF with variants that were not phased by HAPCUT.") val unphasedOutput: PathToVcf,
+  @arg(flag="m", doc="Allow missing fields in the VCF header.") val allowMissingFieldsInVcfHeader: Boolean = true,
```
I would vote for making this hardcoded and not exposing it as a parameter.

nh13 replied:
Done

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+  group=ClpGroups.Phasing
+)
+class HapCutToVcf
+( @arg(flag="v", doc="The original VCF provided to HAPCUT.") val vcf: PathToVcf,
```
Perhaps this is already covered by "provided to HAPCUT", but are there any limitations?  Must this be a single-sample file for example?

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+    val phasedWriter   = makeWriter(output, vcfReader, addPhasedHeaderLines=true)
+    val unphasedWriter = makeWriter(unphasedOutput, vcfReader, addPhasedHeaderLines=false)
+
+    val iterator = new MergingIterator(input, vcfReader, readBackedPhasingFormat)
```
I know it's a private class but `MergingIterator` is a horrible name for this class!  How about `HapCutAndVcfMergingIterator`?

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+      case Right(ctx) => unphasedWriter.add(ctx)
+    }
+
+    CloserUtil.close(vcfReader)
```
`vcfReader.safelyClose()`

nh13 replied:
Awesome! It's in.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+      // get the header lines to skip
+      val headerLinesToSkip = HapCutCallGenotypeInfo.formatHeaderKeys
+        .filter { key => inputHeader.hasFormatLine(key) }
+        .map { key => inputHeader.getFormatHeaderLine(key) }
```
if `inputHeader.getFormatHeaderLine(key)` returns null on a missing key, you could reduce this to:
```scala
HapCutCallGenotypeInfo.formatHeaderKeys.flatMap(key => Option(inputHeader.getFormatHeaderLine(key))
```

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+    val writer: VariantContextWriter = builder.build
+
+    val outHeader = if (addPhasedHeaderLines) {
+      // get the header lines to skip
```
`// get the header lines in the input header that we wish to skip/replace with our own definitions`

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+      val headerLines: util.Set[VCFHeaderLine] = new util.HashSet[VCFHeaderLine](
+        inputHeader.getMetaDataInSortedOrder.filter { line =>
+          if (headerLinesToSkip.contains(line)) {
+            logger.debug(s"Will overwrite header line: '${line.toString}")
```
Is this logging still useful after debugging whatever was going on?  If not, maybe kill it and simplify the whole thing down?

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+      // create the header
+      val header: VCFHeader = new VCFHeader(headerLines, inputHeader.getSampleNamesInOrder)
+      // add the new header lines
+      HapCutCallGenotypeInfo.formatHeaderLines.foreach(header.addMetaDataLine)
```
Why add the standard ones before building the header and these ones after?

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+  }
+}
+
+/** An iterator over all variants in the provided VCFFileReader.  The `input` is the output of HapCut.
```
This line is confusing on first (and second) read.  I would suggest killing `The input is the output of HapCut.` from here, and adding it back as a `@param` doc line to the class.

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+
+  private val sourceIterator = vcfReader.toStream.map { ctx =>
+    // NB: unsets the phase in case it was already set
+    new VariantContextBuilder(ctx).genotypes(new GenotypeBuilder(ctx.getGenotype(0)).phased(false).make()).make()
```
Given that probably 99% of the time you would be using unphased input, do you want to make this conditional on the input genotypes being phased?

nh13 replied:
Suppose you are trying to re-phase the truth data, rather than rely on your own calls, then this makes a lot of sense to unphase, so it is likely "at some point in the future it is 99% of the time" but your point is well taken.

tfenne replied:
No, I get it - that's why I suggested adding an `if (alreadyPhased)` check here and not just removing it.

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+  }
+}
+
+/** Reads phased blocks and variants from a HapCut output file, returning a HapCut call for each input variant call to
```
Something here (and in the usage of HapCutReader) doesn't make sense to me.  The comment says it returns an HapCut call for each input variant, then turns around and says that if HapCut didn't phase a variant `None` is returned because it's not present in the HapCut output.  But this reader class doesn't take in anything _but_ the HapCut output, so how can it know about variants that were not present in the output?

nh13 replied:
The first sentence is wrong, thanks for catching that, done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+
+  def hasNext(): Boolean = {
+    this.block match {
+      case None => this.advance()
```
This can be reduced to 2 cases - drop the `None` and make the last case `_`.

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+  /** Reads in the next block of data */
+  private def advance(): Boolean = {
+    val blockLine = reader.readLine() match {
+      case null => return false
```
Using `return` here is not ideal.  You could move the rest of the method into the `case line` block, and just make the whole method a match statement that yields `true` or `false`.

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+      case null => return false
+      case line => line.trim
+    }
+    if (blockLine == "********") return this.advance()
```
I worry a little bit about hardcoding exactly eight stars - seems like an odd choice from HapCut.  How would you feel about just testing for `line.startsWith('*')`?

nh13 replied:
I think I will change this all around and use `Files.lines` to create an iterator and then it should be trivial to find the next line that starts with "BLOCK".

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+    if (blockLine == "********") return this.advance()
+    val lastOffset = this.block match {
+      case None => 0
+      case Some(b) => b.offset + b.len - 1
```
Could this be a method on the block?  I.e. b.offsetOfLastGenotype?

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+      case None => 0
+      case Some(b) => b.offset + b.len - 1
+    }
+    val currentBlock = HapCutBlockInfo(blockLine, -1)
```
FWIW putting the parsing into the case class objects makes this much harder to follow.

nh13 replied:
Huh, I don't understand why it is harder to follow, since having it in the case class allows me not to be distracted by all the parsing code in this simple method.  What in the case class would you have liked seen here in regards to the parsing code?

nh13 replied:
This is moot, since I have moved it out anyhow, due to the comment about "phaseSet a var" later.

tfenne replied:
I think mostly it’s that I didn’t know the details of the hapcut output format, so it wasn’t clear to me (without detouring) what the values were that were parsed out of the block line, and how those values then influenced the parsing of the subsequent lines.  Now that I have read up on the HapCut format it’s probably a non-issue.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+    // add in the empty calls
+    this.calls = Stream.range(lastOffset+1, currentBlock.len+currentBlock.offset).map { i =>
+      callMap.get(i) match {
+        case None => None
```
Why do you need this match?  You're always returning the thing that was found.

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+        case None => Right(sourceContext)
+        case Some(hapCutCall) =>
+          if (hapCutCall.offset != sourceOffset+1) throw new IllegalStateException("BUG: calls are out of order")
+          val hapCutContext = HapCutCall.toVariantContext(hapCutCall, sampleName)
```
Why isn't this just a method on the `HapCutCall` class?

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+  * @param mecScore the minimum error correction (MEC) score.
+  * @param fragments the number of fragments contributing to variants in this block.
+  */
+private class HapCutBlockInfo private(var phaseSet: Int, val offset: Int, val len: Int, val phased: Int, val span: Int, val mecScore: Float, val fragments: Int)
```
Is `phaseSet` a `var` on purpose?

tfenne replied:
Looking forward, I see why, but I still don't like it.  Options I think would be nicer:
1) Parse the info in the block line outside the case class, and then only create the block when you have all the information
2) Make it a val and use `copy()` to set it once you have the value (and perhaps use `Option[Int]` instead of `Int` with a special value of `-1`.
3) Screw HapCut's block IDs's and just use the offset as the block ids since those are unique and show up in the block line.

nh13 replied:
Done, implemented 2).

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+  val BlockLinePattern = """^BLOCK: offset: (\d+) len: (\d+) phased: (\d+) SPAN: (\d+) MECscore (\d+.\d+) fragments (\d+)$""".r
+
+  def apply(blockLine: String, phaseSet: Int): HapCutBlockInfo = {
+    BlockLinePattern findFirstMatchIn blockLine match {
```
This can be written more idiomatically as the following (I think):
```scala
blockLine match {
  case BlockLinePattern(offset, len, phased, span, mecScore, fragments) =>
    new HapCutBlockInfo(phaseSet, offset.toInt, len.toInt, phased.toInt, span.toInt, mecScore.toFloat, fragments.toInt)
  case _ => throw new IllegalStateException("Could not parse block line: " + blockLine)
}
```

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+/** Information about a variant phased by HapCut.
+  *
+  * @param offset the 1-based offset of the variant relative to the variants given to HapCut.
+  * @param haplotype1 the first allele.
```
hap1Allele?

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+
+  /** Parse a variant line. */
+  def apply(callLine: String, blockOffset: Int): HapCutCall = {
+    val tokens = callLine.split("\t")
```
This may end up being rather slow as it's a regex split on a single character.  You might want to try `htsjdk`'s `StringUtil.split()` instead.

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/phasing/HapCutToVcf.scala:
```diff
@@ -0,0 +1,413 @@
+  }
+
+  /** Converts the HapCut variant representation to a VariantContext. */
+  def toVariantContext(call: HapCutCall, sampleName: String): VariantContext = {
```
I'm going to suggest that the following is an anti-pattern: methods in a companion object that take in, as a parameter, a single instance of the companion class and operate on it.

nh13 replied:
Yup, it is :/

nh13 replied:
I will fix this here and elsewhere, as I am guilty in multiple places.

tfenne commented:
@nh13 A bunch of fairly minor comments, and then one or two more major things I'd like to discuss.

Firstly I think the phasing stuff in VCF is under-specified, and that the GATK docs recommend things that are not ideal.  E.g. I can't find anything in the specification about whether or not the first genotype in a phase set should use `|` or `/` as it's separator.  The example in section `1.1` of the VCF spec shows using `|` in the first genotype in the file(!), while the GATK page you pointed to suggests it should be `/` to differentiate from the prior phasing block.  I would suggest that phasing information in VCF is largely useless without annotated phase sets using the `PS` tag, and so all phased sets should have `PS` set as you are doing, in which case you always know which set/block a genotype belongs to.  Which leads me to saying that all genotypes in a phase set should just use `|`.

Secondly I think the output of a the tool would be better as a single VCF that has genotypes phased where they can be, and unphased where they cannot be.  In addition I would _"phase"_ homozygous genotypes internal to a block.  So if you had say, six genotypes, three of which are phased, two of which are homozygous and one is unphased, you might get the following genotypes in a single file:
```
0|1:PhaseSetA
1|1:PhaseSetA
0|1:PhaseSetA
0/1
0|0:PhaseSetA
1|0:PhaseSetA
```

I could see going either way on the homozygous genotypes, it's just easier to see things if those are also labeled as phased, even though their phase is not really a thing.  But having it all in once file is, I think, more generally useful, and you can always filter to just phased genotypes at runtime by only looking at genotypes with `PS` set.  What do you think?

Lastly, one thing I realized from reading through this and the HapCut doc a little bit is that HapCat is presumably not capable of outputting two parallel "zippered" phase sets.  E.g. imagine if you had intermittent or gappy long reads like the old PacBio strobed reads, that happened to give you phasing of every other het over a region, and another set of long reads that gave you the opposite every-other set.  You have enough information to generate two interleaved phase sets, but I don't think HapCut can do that given the restrictions of it's output format.

nh13 commented:
@tfenne I really appreciate the thorough review, and thinking about the context of the tool and what it tries to do!

1. I have updated the output format to conform to the spec: 
  1. including all phased variants have their alleles separated by `|`, unlike GATK's output.
  2. annotating all phased variants with a PS tag.
2. The output is one file instead of two.
3. I incorporated almost all your suggestions, except one, where I learned about and then used the `java.nio.Files.lines` method, which is quite handy.
4. I agree that HAPCUT doesn't handle the case where you have zipped or overlapping phase sets, which can still be useful in many contexts.  I think this is something to remember after the low hanging fruit has been plucked (phased).  For now, most folks care about the length of the phased blocks, and certain types of switch errors (short, long, isolated), and almost all of these metrics assume no overlapping phased blocks.
5. I tried making the description of the tool more verbose, especially as it relates to the output format.
6. I think the main `execute` method is one of the shortest in any tools we have 👍

tfenne commented on src/main/scala/com/fulcrumgenomics/vcf/HapCutToVcf.scala:
```diff
@@ -0,0 +1,462 @@
+      |
+      |The output of HAPCUT does not include all input variants, but simply those variants that are in phased blocks.
+      |This tool takes the original VCF and the output of HapCut, and produces a VCF containing both the variants that
+      |were phased the variants that were not phased, such that all variants in the original file are in the output.
```
`|were phased _and_ the`

nh13 replied:
Done.

tfenne commented on src/main/scala/com/fulcrumgenomics/vcf/HapCutToVcf.scala:
```diff
@@ -0,0 +1,462 @@
+    val vcfReader      = new VCFFileReader(vcf.toFile, false)
+    val vcfWriter      = makeWriter(output, vcfReader)
+
+    new HapCutAndVcfMergingIterator(input, vcfReader, gatkPhasingFormat).foreach(vcfWriter.add)
```
Very nice!

tfenne commented on src/main/scala/com/fulcrumgenomics/vcf/HapCutToVcf.scala:
```diff
@@ -0,0 +1,462 @@
+  private def formatSourceContext(sourceContext: VariantContext): VariantContext = {
+    val builder = new VariantContextBuilder(sourceContext)
+    if (sourceContext.getGenotypes.exists(_.isPhased)) { // unset the phase if the input has phase set
+      builder.genotypes(sourceContext.getGenotypes().map(g => new GenotypeBuilder(g).phased(false).make()))
```
Isn't this already handled on line 182-184?

nh13 replied:
Will move this out of here, so I only need to make at most one builder.

tfenne commented:
Looks great @nh13.  I'm curious about the decision to set filters on any unphased variants and set `PASS` on any phased variants.  But I think it's fine as is and we could always add an option later to set vs. not set the filters based on phasing status.

tfenne commented:
Oh - one more thought: it might be nice to either open an issue at HapCut, or drop the author a line and let them know about the tool.  I'm sure other HapCut users would like to know about it.

nh13 commented:
@tfenne the details of how GATK represents phase can be seen here: http://gatkforums.broadinstitute.org/gatk/discussion/45/purpose-and-operation-of-read-backed-phasing

In particular, see "More detailed aspects of semantics of phasing in the VCF format".  I am deviating from GATK's output format by not "appending" homozygous variants to the previous phase set.  Note, this output format is only used when the gatk format option is set to true, which is false by default.
